[team-02] iOS: 이슈 상세 화면 및 레이블, 마일스톤 목록 화면 구성

hyeffie opened the pull request:
만사 안녕하세요! 2그룹의 우드, 에피입니다. 구현하느라 구조에 대해 고민하는 시간이 부족했던 것 같아 아쉬움이 많이 남는 한 주 였는데요. 그럼에도 불구하고…! 리뷰 잘 부탁드리겠습니다 😃

## 🧩 주요 작업

이슈 목록 화면에서 네트워크 요청으로 데이터를 받아 항목을 표시하도록 구현했습니다.

- [x]  NetworkManager fetch 메소드 구현

이슈 상세 화면을 구성하고, 서버로 부터 데이터를 전달 받아 표시하도록 구현했습니다.


이슈 목록 화면의 collection view을 구성하는 요소들을 교체했습니다.

- [x]  collection view layout 을 flow layout에서 > compositional layout으로 교체
- [x]  collection view dataSource 를 flow layout에서 > compositional layout으로 교체
- [x]  이슈 닫기, 이슈 삭제를 위한 버튼 추가 (요청 기능 미구현)

이슈 목록 화면에서 사용하는 데이터를 위한 모델 IssueList를 추가하고, 모델 업데이트 로직을 수정했습니다.

- [x]  IssueList 추가 및 view controller 데이터 교체
- [x]  모델 업데이트를 notification으로 처리하도록 구현 변경

레이블 목록 화면과 마일스톤 목록 화면을 구성했습니다. 

- [x]  레이블 목록 화면 구성 및 GET > update 로직 구현
- [x]  마일스톤 목록 화면 구성 GET > update 로직 구현

레이블 추가 화면과 입력 요소를 구성했습니다. (POST 요청 미구현)

- [x]  레이블 추가 화면 구성
- [x]  입력에 따른 화면 업데이트 구현

## 🧩 스크린샷

이슈 목록 화면
![issue-list](https://github.com/codesquad-members-2023/issue-tracker/assets/56967908/650c83a8-1816-4249-a147-611d6327762a)

이슈 목록 화면 swipe action
![issue-list-swipe-action](https://github.com/codesquad-members-2023/issue-tracker/assets/56967908/a52d1b27-38c9-4243-9b4f-ff09ba47e71e)

레이블 목록 화면
![label-list](https://github.com/codesquad-members-2023/issue-tracker/assets/56967908/fd1f0148-ae67-4bff-ab5d-ca5ec9370374)

마일스톤 목록 화면
![mile-list](https://github.com/codesquad-members-2023/issue-tracker/assets/56967908/cee80168-a7f9-4e6b-91a7-84a8516b8d90)

레이블 추가 화면 및 동작
![label-create](https://github.com/codesquad-members-2023/issue-tracker/assets/56967908/178b337f-5e4f-4748-bd5d-cc2a6459cfb8)


## 🧩 기술 키워드

- collection view compositional layout
- collection view diffable data source

## 🧩 고민과 해결 과정

### flow layout vs. compositinal layout

이전에 이슈 목록 화면을 위해 collection view 를 구현할 때에는 화면이 크게 복잡하지 않다고 생각해서 flow layout을 사용했었는데요. 이후에 cell의 sizing 이슈을 해결하는 과정, swipe action을 구현하는 과정에서 collection view + flow layout 의 조합으로 구현에 한계가 있다는 것을 알게 되었습니다.

이 고민에 대해 피드백 시간에 다른 그룹의 해결 과정에 대해 듣게 될 기회가 있었는데, 이때 flow layout을 사용하면서 swipe action을 지원하는 라이브러리를 활용하는 방법, table view로 교체하는 방법 cell height 를 상수로 유지하는 방법 등을 알게 되었지만, 기존 프레임워크를 최대한 활용하고, self-sizing하도록 구현할 수 있는 방법으로 해결해보고 싶다고 생각했던 것 같습니다.

그래서 compositional layout의 list configuration을 활용해서 원하는 리스트 UI를 구성하고, table view 처럼 swipe action을 추가할 수 있고, cell의 크기를 고정하지 않고, content에 따라 자동으로 크기를 결정할 수 있도록 구현하기로 결정하게 되었습니다.

## 🧩 질문

### 목록 화면 공통 요소 리팩토링

이번 스프린트 목표 중에 하나가 세 개의 탭에서 공통적으로 사용되고 있는 목록 화면을 구현하는 공통 코드를 만들어 쉽게 구현하고 재사용성을 높이자! 였는데요. DiffableDataSource를 사용하다 보니 목록 화면 각각에 필요한 data source의 타입이 달라지게 되고, 결국엔 원하는 만큼 재사용할 구현을 분리하는 것이 어렵다고 느꼈던 것 같습니다.

혹시 DiffableDataSource를 사용하면서 공통 화면에 대한 구현을 재사용하고 싶다면

- ViewController의 동작을 프로토콜로 추상화하는 방법
- ListingViewController라는 슈퍼 클래스를 구현하고, 이를 서브클래싱 하는 방법
- data source를 view controller에서 분리하는 또 다른 방법

중에 바람직한 방법이 있다고 생각하시는지 만사의 의견이 궁금합니다…! 또는 제시한 것들 이외에 자주 사용되는 방법이나 패턴이 있다면 추천을 부탁드려도 될까요?

hyeffie committed:
feat #68: IssueListTableViewCell configure(issue:) 메소드 구현

- issue 를 전달 받아 셀 content를 설정하는 configure(issue:) 메소드 구현

hyeffie committed:
feat #68: collection view layout 교체

- flow > compositional layout
- list configuration 적용

hyeffie committed:
feat #68: collection view data source 교체

- UICollectionViewDataSource (self) > UICollectionViewDiffableDataSource
- LoadCollectionViewCell xib 파일의 reuse identifier 제거
- IssueListDTO.Issue 에 Hashable 채택 및 구현

hyeffie committed:
feat #68: cell - label stack view hidden 구현

- 이슈 label 배열이 비었을 때 stack view를 hidden 처리하는 로직 구현
- prepare for reuse 코드에서 show 처리

hyeffie committed:
feat #68: issue cell contextual action 버튼 추가

- layout configuration 의 trailingSwipeActionsConfigurationProvider 에서 닫기, 삭제 action 생성 및 설정하는 provider 구현 추가

hyeffie committed:
fix #77: label stack view distribution style 변경

- distribution을 fill > fill proportionally 로 변경

hyeffie committed:
feat #69: IssueList 구현

DTO를 대체할 모델로서 IssueList 구현

hyeffie committed:
fix #69: DTO.Issue.milestoneName nullable 처리

- DB 변경사항 반영
- 임시 force upwrapping 사용

hyeffie committed:
feat #69: IssueListVC에 Notification 적용

- Model: 모델 업데이트 > post noti 
- VC: addObserver > observed > update UI

hyeffie committed:
chore #69: CellIdentifiable 추가 및 코드 정리

hyeffie committed:
feat #70: ListingItemFactory 추가

- DTO > 모델 converting 용 factory 추가

hyeffie committed:
feat #76: SwipeAction 타입 추가 및 IssueListVC 구현 추가

- SwipeAction 추가
- IssueListVC collection view trailing action provider 구현

hyeffie committed:
feat #70: LabelListVC 추가

- LabelListVC ViewController 추가
- tab bar 레이블 탭 추가

hyeffie committed:
feat #70: LabelListVC collection view 구현

- 관련 타입 추가
- 코드로 VC 생성 및 연결
- 하드 코딩 데이터로 동작 확인
- 서버 DB 변경 사항 반영 (issue isOpen > open)

hyeffie committed:
feat #76: swipe action 추가

hyeffie committed:
feat #71: 마일스톤 탭 추가

- Milestone List 추가
- MilestoneListViewController  구현
- tab bar 에 마일스톤 챕 추가

hyeffie committed:
feat #71: MilestoneListVC collection view 추가

- 셀 디자인 및 타입 추가
- MilestoneList 모델 일분 속성 변경
- full swipe 속성 변경

hyeffie committed:
feat #71: 내 계정 탭 추가

hyeffie committed:
feat #70: label list GET 구현

- label list fetch 메소드 구현
- 프로젝트 그룹 정리

hyeffie committed:
feat #70: label stack view empty 로직 추가

hyeffie committed:
feat #71: milestone list GET 구현

- GET 요청 구현
- MilestoneDTO 추가
- 모델 업데이트 반영 구현

hyeffie committed:
chore #71: 프로젝트 및 코드 정리

hyeffie committed:
feat #108: 레이블 추가 화면 구성

- 목록 화면에 버튼 추가
- table view + static cell 로 추가 화면 구성

hyeffie committed:
feat #108: 속성 입력 기능 구현

- 속성 및 outlet 추가
- UI 설정 메소드 구현
- 입력 관련 로직 구현
- UITextField 확장 구현

ITzombietux commented on ios/IssueTracker/IssueTracker/Common/CellIdentifiable.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+protocol CellIdentifiable {
+   static var cellId: String { get set }
+}
```
이 프로토콜은 어떤 역할인지 여쭤보고 싶군요! 네이밍은 역할을 드러내고 있지만 cellId는 set도 가능하다는 점에서 궁금증이 생깁니다🤔

hyeffie replied:
```swift
   private func createLoadCellRegisteration() -> UICollectionView.CellRegistration<LoadCollectionViewCell, ItemType> {
      let loadCellNib = UINib(nibName: LoadCell.cellId, bundle: nil)
      return .init(cellNib: loadCellNib, handler: { _, _, _ in })
   }
   
   private func createCellRegisteration<Cell: CellIdentifiable, Item>(cellCompletion: @escaping UICollectionView.CellRegistration<Cell, Item>.Handler) -> UICollectionView.CellRegistration<Cell, Item>  {
      let cellNib = UINib(nibName: Cell.cellId, bundle: nil)
      return .init(cellNib: cellNib, handler: cellCompletion)
   }
```

위 메소드처럼 Cell 타입의 cell id를 통해 바로 register 하는 코드를 작성하고 싶었는데, cell의 content를 설정하는 로직에 대응하는 방법을 찾지 못해서 고민하고 있었던 중이었습니다. 저도 현재 구현은 불필요하다고 생각하고, 다른 구현으로 프로토콜을 작성해 볼 생각입니다!

ITzombietux commented on ios/IssueTracker/IssueTracker/IssueTab/IssueListFilter/FilterListViewController.swift:
```diff
@@ -98,7 +98,7 @@ extension FilterListViewController: UICollectionViewDataSource {
             filterElement = useCase.labelList[indexPath.row].labelName
          default:
-            filterElement = useCase.milestoneList[indexPath.row].milestoneName
+            filterElement = useCase.milestoneList[indexPath.row].milestoneName!
```
가능하면 강제 언래핑으로 인한 에러발생은 막는게 좋겠죠?? 👍

hyeffie replied:
네 수정하도록 하겠습니다 !

ITzombietux commented on ios/IssueTracker/IssueTracker/IssueTab/List/IssueListViewController.swift:
```diff
@@ -0,0 +1,196 @@
+   var observers: [NSObjectProtocol] = []
+   
+   @IBOutlet weak var collectionView: UICollectionView!
+   private var dataSource: DataSource!
```
dataSource가 초기화되지 않을 일은 없을까요?! ㅎㅎ

hyeffie replied:
스냅샷을 적용하지 않더라도, 그러니까 collection view의 content는 초기화되지 않더라도 최소한 어떤 형태의 data source 가 될 것인지 결정해주어야 컬렉션 뷰를 사용할 수 있지 않을까 해서 내린 결정이었는데,,, data source를 초기화하지 않은 상황을 생각해보지 못했던 것 같아요. 좀 더 고민해보겠습니다!

ITzombietux replied:
강제 언래핑을 사용할 땐 명확한 근거가 있어야 좋다는 생각이 들어 의견을 남기게 됐어요!! ㅎㅎㅎ

ITzombietux commented on ios/IssueTracker/IssueTracker/MilestoneTab/List/MilestoneListCell.swift:
```diff
@@ -0,0 +1,66 @@
+      nameLabel.text = milestone.name
+      progressLabel.text = "\(milestone.progress)%"
+      
+      if let desc = milestone.description, desc.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false {
```
desc보다는 풀네임이 좋겠죠! 추가로 두 가지 조건식이지만 코드가 길어져 가독성이 떨어지면 조건을 분리해서 속성으로 만들어 버리는 방법도 있습니다!!

hyeffie replied:
줄여서 지은 네이밍이 많은데 신경써야겠네요....! 저도 lint line length 룰 때문에 라인이 길어지는 걸 어떻게 처리하면 좋을지 고민이었는데 팁 감사합니다! 수정해보겠습니다 😃

ITzombietux commented on ios/IssueTracker/IssueTracker/MilestoneTab/List/MilestoneListCell.swift:
```diff
@@ -0,0 +1,66 @@
+         descriptionLabel.isHidden = true
+      }
+      
+      if let completedAt = milestone.completedAt, completedAt.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false {
+         completedAtLabel.text = milestone.completedAt
+      } else {
+         dueContainer.isHidden = true
+      }
```
if else 와 같은 condition이 UI에 존재하면 테스트하기 어려울 수 있기 때문에 의사결정이 필요한 부분이 있죠! completedAt이 테스트가 돼야 하는 기능, 중요한 기능인지를 따져봐야겠죠!!

hyeffie replied:
이 부분이 셀의 content를 설정하는 로직인데, 항상 조건에 따라 다른 UI 를 보여주고자 할 때 반복적으로 이런 방식의 구현을 해왔었거든요. 테스트 관점으로는 생각을 안해봤는데 고민해 볼 필요가 있다는 생각이 드는 것 같아요. 로직을 분리하는 방식 등으로 개선해보겠습니다.

ITzombietux replied:
Cell에 Label 모양이 결정된 상태를 넘겨주면 좋지 않을까 싶어요!!

ITzombietux commented on ios/IssueTracker/IssueTracker/MilestoneTab/List/MilestoneListViewController.swift:
```diff
@@ -0,0 +1,140 @@
+   }
+   
+   func setNetworkManager() {
+      networkManager = NetworkManager(session: URLSession.shared)
```
networkManager와 같이 ViewController와 밀접하게 협력하는 객체는 테스트가 되어야 할 객체라고 볼 수 있는데요! 그렇다면 setNetworkManager와 같은 동작이 아닌 init에 넣으면 좀 더 유연하게 테스트를 할 수 있는 환경이 제공되지 않을까 생각해봅니다 ㅎㅎ

hyeffie replied:
뭔가 언젠가 부터 당연히 viewDidLoad에서 모든 설정을 담당하도록 구현하고 있었네요. 😅 주입하는 방법도 충분히 고려해보겠습니다!!

ITzombietux commented on ios/IssueTracker/IssueTracker/MilestoneTab/List/MilestoneListViewController.swift:
```diff
@@ -0,0 +1,140 @@
+      collectionView.translatesAutoresizingMaskIntoConstraints = false
+      view.addSubview(collectionView)
+      
+      NSLayoutConstraint.activate([
+         collectionView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor),
+         collectionView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor),
+         collectionView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
+         collectionView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor)
+      ])
```
메서드는 setCollectionView이고 setCollectionViewLayout 메서드가 또 있는데 책임이 분산이 됐군요! ㅎㅎㅎ

ITzombietux commented on ios/IssueTracker/IssueTracker/MilestoneTab/List/MilestoneListViewController.swift:
```diff
@@ -0,0 +1,140 @@
+
+extension MilestoneListViewController {
+   func fetchMilestones(cellCompletion: (() -> Void)? = nil) {
+      networkManager?.requestMilestoneList() { [weak self] dto in
+         cellCompletion?()
+         if dto.milestoneList.count > 0 {
+            let milestones = ListingItemFactory.MilestoneTab.makeMilestoneList(with: dto)
+            self?.list.emptyList()
+            self?.list.add(milestones: milestones)
+         }
```
추후에 UseCase가 확실히 나뉘어지면 이런 로직이 옮겨지면 좋겠네요!!

hyeffie replied:
맞습니다....! 이 얘기를 들으니까 use case 도입하고 싶었던 마음이 다시 자극받는 것 같네요.. 🤤 다른 구현을 어서 마치고 꼭 개선해보겠습니다 !!

ITzombietux commented on ios/IssueTracker/IssueTracker/Model/IssueList.swift:
```diff
@@ -0,0 +1,178 @@
+import Foundation
+
+class IssueList {
+   class Issue: Hashable {
```
IssueList안에 Issue를 만든 특별한 이유라도 있을까요???🤔두 객체의 역할이 명확하게 구분이 되지 않을까 싶기도 한데요 !

hyeffie replied:
저도 처음 모델로 사용될 클래스를 nested로 구현한다는 사실이 좀 이상하게 여겨지기도 했는데, 외부에 Issue를 선언하면 다른 Issue와 중복이 생겨서 결국에는 nested로 구현하기로 결정했었는데요. 이름을 바꾸면 된다는..ㅋㅋ  놀라운 방법을 알아내서 반영해보려고 합니다. 감사합니다!

ITzombietux reviewed (approved):
한주간 많은 부분을 구현해주셨네요 수고하셨습니다! 리뷰를 하며 같이 생각해봤으면 하는 부분을 남겨볼게요!

1️⃣ 구현한 부분 + 핵심 코드베이스 강조
추가된 코드가 많아질수록 모든 코드베이스를 같은 중요도를 갖고 리뷰할 수 없죠! 그렇기 때문에 많은 코드 중에서도 "핵심 코드베이스가 어떤 부분이고 히스토리는 이러하다." 는 공유가 있으면 훨씬 더 효율적인 소통이 될 수 있을 거 같습니다☺️

2️⃣한 걸음씩 완료하며 얻는 성취
일주일의 시간이지만 한 걸음씩 꾸준히 잘 나아간 결과가 보여서 너무 좋습니다! 시간에 쫓겨 아쉬운 구현이야 분명히 있겠지만 일주일의 완료가 있기에 다음 완료를 기대할 수 있는 것처럼 남은 기간도 하나씩 완료해나가며 파이팅 하시기를 바랄게요!!

질문에 대한 제 의견은,

3. data source를 view controller에서 분리하는 또 다른 방법

data source를 잘 추상화해보면 재사용이 충분히 가능하지 않을까? 하는 생각입니다!! 다른 방법도 고민해보면 좋긴 하겠네요!!

wood0203 committed:
feat #67: 상세 화면 Storyboard 생성

- 네비게이션 바 우상단 수정 버튼 추가
- 툴바 표시를 위해 하단 탭바 숨김 처리
- 코멘트 입력과 이동에 필요한 하단 툴바 추가

wood0203 committed:
feat #67: 이슈 목록 > 이슈 상세 화면 이동 구현

- 셀 선택시 해당 이슈 상세페이지로 이동
- 뒤로가기위해 push로 이동하도록 구현

wood0203 committed:
feat #67: 이슈 상세 페이지 데이터 로드

- 이슈 상세 DTO 모델 생성
- 이슈 상세 데이터 받아오는 메소드 구현
- 이슈 목록 > 이슈 상세화면으로 이슈 번호 전송 구현

wood0203 committed:
feat #67: 이슈 상세 유스케이스 생성

- 데이터 불러오는 메소드 추가

wood0203 committed:
feat #67: 콜렉션 뷰 추가

- 사용할 커스텀 헤더 등록
- UI 테스트를 위해 Dummy Cell 등록

wood0203 committed:
feat #67: 유스케이스 > 헤더 셀 DTO 생성

- 헤더 셀에 필요한 데이터를 담는 데이터 구조체 추가

wood0203 committed:
feat #67: 이슈 상세정보 헤더 표시

- DTO 이용하여 헤더 출력
- 16진수 > UIColor 생성 팩토리 추가
- 생성/닫힘 시간 격차 계산 객체 추가
- ViewController 기본 UI 설정 메소드 캡슐화

wood0203 committed:
feat #67: ViewController 옵저버 등록

- 유스케이스 데이터 로드 완료시 콜렉션 뷰 재로드
- 옵저버 등록 메소드 추가
- IssueDeatilDTO 확장하여 알림 이름 프로퍼티 추가

wood0203 committed:
fix #67: 시간 격차 생성 객체 변경

- 소수점 삭제하기 위해 계산하는 타입 Double > Int 변경

wood0203 committed:
feat #67: 콜렉션 뷰 배경색 적용

- 셀간 간격 확인을 위해 적용

wood0203 commented:
만사 너무 늦은 PR 죄송합니다..
이번주에 컨디션이 많이 안 좋았어서 구현이 적어 늦게나마 결과물을 내서 보내드려는 욕심을 냈더니,,
다음 PR은 시간을 꼭 준수하도록 하겠습니다😭

## 🧩 주요 작업
- 이슈 목록에서 이슈 선택시 상세 페이지로 이동하고, 이슈 정보를 표시합니다.
- 유스케이스를 분리했습니다.
- 이슈 정보를 표시하는데 필요한 데이터들을 생성해주는 팩토리(or 객체)를 구현하였습니다.
   - ex) LastTimeGenerator / UIColorFactory
   - 뷰 컨트롤러에서 계산하지 않고, 역할을 분리시키고 로직을 숨기고자 의도했습니다.
- 옵저버 패턴을 사용하여 비동기 데이터 로드 완료시 정보를 표시합니다.

## 🧩 스크린샷
![Simulator Screen Recording - iPhone 14 Pro - 2023-05-27 at 14 40 26](https://github.com/codesquad-members-2023/issue-tracker/assets/84387335/08c70ceb-d7ba-4482-80dc-8ec9597f6b30)

## 🧩 기술 키워드
- UseCase
- Observer Pattern

## 🧩 고민과 해결 과정
- ViewController에서 데이터를 받는 로직을 숨길 수는 없을까?
저번주 작업간 후회?되었던건, 데이터 로드 비즈니스 로직을 뷰컨트롤러가 알고 있다는 점이였습니다.
그래서 유스케이스를 분리하였고, 뷰 컨트롤러는 유스케이스의 Load메소드 하나만 호출하면 로드가 완료되도록 했습니다.

## 🧩 질문
- 이슈 상세화면에서 데이터를 Notification에 담아서 전송하는게 더 효율적일지 궁금합니다.
현재 사용중인 옵저버 방식은 다음과 같습니다.
1. UseCase에서 비동기 데이터 로드 완료시 Noti 전송
2. ViewController에서 UseCase에 필요한 데이터 요청   

그러나 이슈 상세화면의 경우 데이터 수가 적고,
ViewController에서 UseCase에 다시 요청하는 과정을 줄인다면 더 효율적이지 않을까 생각해봤습니다😂

hyeffie committed:
chore: 코드 정리

- coding style 적용
- lint disabled rule 추가 line length 수정

hyeffie committed:
chore: 프로젝트 구조 정리

hyeffie committed:
chore: IssueListVC code base 로 변경

- storyboard > code
- 기타 코드 정리
- CellIdentifiable 제거

hyeffie commented:
@ITzombietux 
3주간 프로젝트를 진행하면서 구현과 구현을 위한 학습, 그리고 더 좋은 구조에 대한 충분한 고민 사이에서 계속해서 딜레마를 겪어 왔었고, 할 수 있는 것을 우선 해나자는 생각으로 달렸는데 그 과정을 알아주시는 듯한 말씀이 힘이 됩니다. 다음 스프린트에서 목표를 완수해야겠다는 마음을 한 번 더 다지게 되는 한편, 리뷰의 내용을 더 담아낼 수 있도록 이후에 꼭 개선해봐야겠다. 그런 생각을 하게 되는 것 같아요.
의견 주신 방법대로 재사용, 그리고 역할 분리에 대해서는 우선적으로 적용을 해보려고 합니다. 그리고 다음 PR에서는 말씀하신 것처럼 핵심 구현에 대해서 좀 더 효율적으로 공유해보도록 하겠습니다.
빠른 리뷰 감사합니다! 좋은 주말 되세요 🙂

ITzombietux commented on ios/IssueTracker/IssueTracker/Common/CellIdentifiable.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+//protocol CellIdentifiable {
+//   static var cellId: String { get set }
+//}
```
사용하지 않을 때는 과감하게 삭제하는 것도 좋은 연습이겠죠?ㅎㅎ

ITzombietux commented on ios/IssueTracker/IssueTracker/Common/LastTimeGenerator.swift:
```diff
@@ -0,0 +1,44 @@
+      return component
+   }()
+   
+   func calculateLastTime(past: String) -> String {
```
dateFormatter를 사용하여 시간을 계산하는 동작을 갖는 객체라면 static하게 사용하는 방법도 고려해보시면 좋을 거 같습니다 ㅎㅎ

ITzombietux commented on ios/IssueTracker/IssueTracker/Common/ListingItemFactory.swift:
```diff
@@ -0,0 +1,67 @@
+
+import Foundation
+
+struct ListingItemFactory {
```
이 Factory에 대해서 한번 고찰을 해볼 필요가 있는데요! 뭔가 마법같은 단어로 사용이 가능하지만 그렇기 때문에 더 잘 사용해야 하는 단어가 Factory인 거 같습니다. 그렇기 때문에 어떤 경우에 사용하면 좋은지를 잘 정리하는 게 좋을 거 같습니다. 저는 Factory를 사용하기 좋은 상황은 특정 동작이 반복될 때 중복을 제거하는 용도로 Factory를 사용하는데요. 약간 더 추상화되겠죠? IssueTab과 LabelTab의 중복을 제거하면 더 좋은 Factory가 될 거 같네요! 그렇게 된다면 ListingItemFactory안에 Nested한 sturct도 제거할 수 있겠죠!

hyeffie replied:
오 무슨 말인지 이해한 것 같아요! 반영해보겠습니다. 감사합니다!

ITzombietux commented on ios/IssueTracker/IssueTracker/Common/SwipeAction.swift:
```diff
@@ -0,0 +1,52 @@
+
+import UIKit
+
+enum SwiptAction {
```
오타를 발견했네요!!😁😁😁😁😁

hyeffie replied:
...!! 어서 고치겠습니다!

ITzombietux commented on ios/IssueTracker/IssueTracker/Common/UIColorFactory.swift:
```diff
@@ -0,0 +1,26 @@
+
+import UIKit
+
+struct UIColorFactory {
```
color.swift 파일과 약간 겹치는 역할인 것 같은데요?! ㅎㅎ

wood0203 replied:
감사합니다 겹치는 부분 삭제하도록 하겠습니다!

ITzombietux commented on ios/IssueTracker/IssueTracker/IssueTab/Detail/IssueDetailViewController.swift:
```diff
@@ -0,0 +1,158 @@
+      
+      let headerData = issueDetailUseCase.sendHeaderData()
+      
+      let time = lastTimeGenerator.calculateLastTime(past: headerData.editTime)
+      header.editTime.text = "\(time) 전, \(headerData.userName)님이 작성했습니다."
+      header.title.text = headerData.title
+      header.number.text = "#\(headerData.number)"
+      
+      switch headerData.status {
+      case false:
+         header.status.text = "열린 이슈"
+         header.status.backgroundColor = uiColorFactory.make(hexColor: openColor)
+      case true:
+         header.status.text = "닫힌 이슈"
+         header.status.backgroundColor = uiColorFactory.make(hexColor: closeColor)
+      }
```
이 로직은 VC보다는 View에 있다면 더 좋을 수 있을 거 같습니다 ㅎㅎ

wood0203 replied:
감사합니다!! 역할을 좀 분리시키도록 반영해보겠습니다!

ITzombietux commented on ios/IssueTracker/IssueTracker/IssueTab/List/IssueListViewController.swift:
```diff
@@ -0,0 +1,246 @@
+
+extension IssueListViewController {
+   private func addObservers() {
+      let noti = NotificationCenter.default.addObserver(
```
이틀 전에 동일한 코멘트가 있었는데, noti보다는 fullname으로 적는 게 좋을 거 같네요!

ITzombietux reviewed (approved):
주말인데 너무나 수고하셨습니다!👍
다음 스프린트에서는 ViewController와 View의 책임을 조금만 더 신경써서 코드를 작성해보시면(물론 머리는 많이 아플 수 있지만,,,🥲) 너무 좋을 거 같네요!!

ITzombietux commented:
> @ITzombietux
> 3주간 프로젝트를 진행하면서 구현과 구현을 위한 학습, 그리고 더 좋은 구조에 대한 충분한 고민 사이에서 계속해서 딜레마를 겪어 왔었고, 할 수 있는 것을 우선 해나자는 생각으로 달렸는데 그 과정을 알아주시는 듯한 말씀이 힘이 됩니다. 다음 스프린트에서 목표를 완수해야겠다는 마음을 한 번 더 다지게 되는 한편, 리뷰의 내용을 더 담아낼 수 있도록 이후에 꼭 개선해봐야겠다. 그런 생각을 하게 되는 것 같아요.
> 의견 주신 방법대로 재사용, 그리고 역할 분리에 대해서는 우선적으로 적용을 해보려고 합니다. 그리고 다음 PR에서는 말씀하신 것처럼 핵심 구현에 대해서 좀 더 효율적으로 공유해보도록 하겠습니다.
> 빠른 리뷰 감사합니다! 좋은 주말 되세요 🙂

완료를 향해 달려가며 구현, 피드백 반영 모든 것을 포함해서 두 분의 의사결정을 통해 어떤 것에 힘을 줄지 잘 결정하시기를 바랍니다! 모든 로직을 재사용할 필요도 없고 모든 코드의 역할 분리를 할 필요도 없다고 생각합니다. 두 분이 생각할 때 앱의 가장 중요한 부분부터 차근차근 완료해보시기를 바래요!!

wood0203 commented:
주말인데도 불구하고 꼼꼼히 리뷰 해주셔서 너무 감사합니다! 다음 PR은 좀 더 초기설정한 목표를 완성하도록하고 말씀해주신 역할과 책임을 잘 생각해보도록 노력해보겠습니다. 또 PR에 더 담아내려고 노력해보겠습니다!
